### PR TITLE
[Draft] Adding Multi device iterator Serialization and De-serialization

### PR DESCRIPTION
**Pull request by sboshin**
Context: 
The idea is to be able to save and restore iterators (The iterators that can be saved) for multi worker training. This will allow for training to be resumed mid-epoch instead. This reduces wasted compute cycles when restarting training jobs.
With this change the Backup and restore keras callback can include the iterator (if supported), and actually restore the keras model to a state mid epoch also.

Description:
Adding a serialization and de-serialization to multi device iterator.
Added Save and Restore in both the Multi Device Resource and the Multi Device Buffer.

Design.
New operators added for Multi Device Iterators to serialize and de-serialize
Multi device buffer will save its buffer and host_iterator.
Multi device Resource will restore by creating a new buffer because the host iterator is const.
It will then call the multi device buffer to restore its own buffer.

Test.
Using a Range Dataset(10)
Save after 0-5, then restore in a new session, and verify 6-9

### CODE REVIEW

**Comment by gbaned** [comment]
@sboshin  This PR is in draft, any update on this? Please. Thanks!

**Review by jsimsa — CHANGES_REQUESTED** [review]
Thank you for your contribution. I have some high-level comments:

1) This is a non-trivial change that should be coupled with a document that provides a high-level overview of your design for checkpointing of Multi-Device Iterators. We should first reach a consensus on the document before continuing review of the PR.

In particular, the design should address how will checkpointing ascertain correctness in the presence of multiple input pipelines (one on the host and one on each fo the devices) executing in parallel. The only test you have added does not use `prefetch` on the device input pipeline, which makes me wonder whether the design and implementation handles that case correctly.

2) When making code changes, please avoid re-formatting code that you are not modifying. It makes the code review process less efficient.

4) The change should avoid code duplication. For instance, it is not immediately clear why `IteratorVariantSerializer` needs to be duplicated in multi_device_iterator_ops.cc.

3) We will want to test the functionality in both TF 1 and TF 2. The current tests only cover TF 1.

**Comment by gbaned** [comment]
@sboshin  Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 16 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@sboshin Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
